modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStorage.scala

```diff
@@ -0,0 +1,222 @@
+    lastKeyAndArtifact: Option[(Key, Artifact)] = none[(Key, Artifact)]
+  ): F[ConsensusStorage[F, Event, Key, Artifact]] =
+    for {
+      stateUpdateSemaphore <- Semaphore[F](1)
+      lastKeyAndArtifactR <- Ref.of(lastKeyAndArtifact)
+      eventsR <- MapRef.ofConcurrentHashMap[F, PeerId, List[(Ordinal, Event)]]()
+      statesR <- MapRef.ofConcurrentHashMap[F, Key, ConsensusState[Key, Artifact]]()
+      resourcesR <- MapRef.ofConcurrentHashMap[F, Key, ConsensusResources[Artifact]]()
+    } yield make(stateUpdateSemaphore, lastKeyAndArtifactR, eventsR, statesR, resourcesR)
+
+  def make[F[_]: Async: KryoSerializer, Event, Key: Show: Next: Eq, Artifact <: AnyRef: Show: Eq](
+    stateUpdateSemaphore: Semaphore[F],
+    lastKeyAndArtifactR: Ref[F, Option[(Key, Artifact)]],
+    eventsR: MapRef[F, PeerId, Option[List[(Ordinal, Event)]]],
+    statesR: MapRef[F, Key, Option[ConsensusState[Key, Artifact]]],
+    resourcesR: MapRef[F, Key, Option[ConsensusResources[Artifact]]]
+  ): ConsensusStorage[F, Event, Key, Artifact] =
+    new ConsensusStorage[F, Event, Key, Artifact] {
+
+      private val logger = Slf4jLogger.getLogger[F]
+
+      def getState(key: Key): F[Option[ConsensusState[Key, Artifact]]] =
+        statesR(key).get
+
+      def getResources(key: Key): F[Option[ConsensusResources[Artifact]]] =
+        resourcesR(key).get
+
+      def updateState(key: Key)(stateUpdateFn: StateUpdateFn): F[Option[ConsensusState[Key, Artifact]]] =
+        stateUpdateSemaphore.permit.use { _ =>
+          for {
+            (maybeState, setter) <- statesR(key).access
+            result <- stateUpdateFn(maybeState, setter)
+            maybeNewState <- result.flatTraverse {
+              case (true, newState) =>
+                logger.trace(s"Consensus state for key ${key.show} transitioned to ${newState.show}") >>
+                  Applicative[F].pure(newState.some)
+              case (false, newState) =>
+                /** This should never occur as the access to the `ConsensusState` is protected by the semaphore */
+                logger.error(s"Consensus state for key ${key.show} failed to transition to ${newState.show}") >>
+                  Applicative[F].pure(none[ConsensusState[Key, Artifact]])
+            }
+          } yield maybeNewState
+        }
+
+      def setLastKeyAndArtifact(value: Option[(Key, Artifact)]): F[Unit] = lastKeyAndArtifactR.set(value)
+
+      def getLastKeyAndArtifact: F[Option[(Key, Artifact)]] = lastKeyAndArtifactR.get
+
+      def tryUpdateLastKeyAndArtifactWithCleanup(oldValue: (Key, Artifact), newValue: (Key, Artifact)): F[Boolean] =
+        lastKeyAndArtifactR.modify { maybeValue =>
+          if (maybeValue === oldValue.some)
+            (newValue.some, true)
+          else
+            (maybeValue, false)
+        }.flatTap(_ => cleanupStateAndResource(oldValue._1))
+
+      private def cleanupStateAndResource(key: Key): F[Unit] =
```

Comment:
Is it compatible with `access` update used in `updateState` function. I mean no race condition possible?